### ISSUE TITLE
kv: price the boundary-layer K8V8 promotion — the iso four pay 1.32x bytes for a bf16 fidelity floor nobody has costed

### ISSUE THREAD

**Issue by Pushkinist** [Pushkinist/rMLX#443]
## Summary

Eight store-materialising KV codecs have their head/tail **boundary layers
promoted upward into `K8V8`**, which is itself inert — a `K8V8` layer holds two
full bf16 buffers and nothing else, 16 bits per value. For the four
`SideStore::IsoRing` codecs that is a **1.32× byte regression** on those layers,
paid deliberately to keep a bf16 fidelity floor. The trade is now documented and
correct. This issue exists so it stays **revisitable** rather than becoming
folklore.

## The mechanism

`boundary_floor` (`crates/rmlx-models/src/kv_cache/mod.rs:279`) applies a
quality floor at 8 bits to the standard 2 head + 8 tail layers. A base whose
widths are *parameters* (`Mixed`, `RotK`) is raised inside its own family — same
store, same group geometry, same K rotation, both axes to 8 bits — and delivers
the floor from bytes the layer already spends (8.50 bits per value at group 64).
A base whose width is *baked into its variant* has no 8-bit form of its own
family to raise to and falls back to `K8V8`.

Ten codecs materialise a packed store. Two are parametric. **Eight take the
`K8V8` fallback**, and for those eight the floor is paid for, not free:

| group | codecs | store bits/value | `K8V8` | effect on boundary layers |
|---|---|---|---|---|
| `SideStore::IsoRing` | `Iso3Sym`, `Iso4Sym`, `IsoKOnly3`, `IsoKOnly4` | **12.125** | 16.00 | **1.32× byte regression** |
| `SideStore::Rotor` | `Rotor3Sym`, `Rotor4Sym`, `RotorKOnly3`, `RotorKOnly4` | 16.25 | 16.00 | byte-neutral to slightly favourable |

The iso row is the one that costs. An SO(4)-rotated 3-/4-bit ring has no 8-bit
form, and the alternative to paying for the floor is not having one — so the
trade as taken is defensible. It is still a trade, and it is the reason iso's
whole-cache ratio does not reach its store ratio.

## Measured — the gap the trade explains

`rmlx serve` at `pre-release/kv-evidence` @ `8fd237a8`, 928-token prompt,
temp=0, `kv_cache_bytes` from the server's N16 event, ×`none`:

| model | `iso3_sym` | `k_iso3` | store ratio would predict |
|---|---|---|---|
| Ternary-Bonsai-8B | 0.962× | 0.980× | 0.758× |
| Ternary-Bonsai-27B | 0.982× | 0.991× | 0.758× |
| Qwen3.8-27B | 0.971× | 0.986× | 0.758× |
| gemma-4-e2b | 0.876× | 0.937× | 0.758× |
| gemma-4-12B | 0.980× | 0.996× | 0.758× |

`12.125 / 16.0 = 0.758` is what the iso store alone would deliver. The observed
0.876–0.996 is that store on the non-boundary layers plus **bf16 on the 10
promoted ones**. On Bonsai-8B, 10 of 36 layers are promoted — over a quarter of
the stack held at 16.0 bits per value — and that is the whole of the difference
between "iso is a 24% memory win" and "iso is a 2–4% memory win".

## Why file it rather than leave it in the docstring

The docstring argues the trade correctly and `parametric_boundary_promotion_stays_under_bf16`
pins it. What neither records is that the trade **was never priced against its
alternatives**, and it is the single largest term in the iso family's
whole-cache rate. Three alternatives exist and none has been measured:

1. **A narrower floor.** The floor is 8 bits because that is the widest width
   `validate_mixed_side` accepts. For a ring-backed codec the equivalent
   fidelity floor might be its own 4-bit form (`Iso4Sym` as the target for
   `Iso3Sym`), which is above the base and *below* bf16. Costs nothing to
   evaluate; the ring already has both widths and they are byte-identical.
2. **Fewer promoted layers.** `LAYER_ADAPTIVE_TAIL_N` = 8 tail + 2 head is
   inherited, not derived from an iso-family measurement. If 4 tail layers hold
   the quality, the byte penalty halves.
3. **No floor for this family at all.** Measure PPL / greedy divergence of an
   un-floored iso cache against a floored one. If the floor is buying nothing on
   this family, it is a pure 1.32× tax on a quarter of the stack.

## Asks

* Measure (1) and (3) on ≥2 architectures (`gemma-4-e2b` for `kv_h == 1`
  shared-KV, `Ternary-Bonsai-8B` for `kv_h > 1` dense — the standard pair), at a
  context sweep, on PPL **and** greedy-token divergence, not bytes alone.
* Record the result in `docs/KV_QUANT.md` § Layer-adaptive overrides beside the
  existing trade, whichever way it goes. A negative result is the useful outcome
  here: it converts a defensible-looking default into a measured one.
* If the floor stays as is, say in the docstring that the alternatives were
  priced and name the numbers — that is what makes it revisitable instead of
  inherited.

## Not in scope

This is not a proposal to remove the floor. The floor exists because a boundary
layer at 3 bits measurably hurts, and the campaign that added it recorded 37–91%
turbo2 recovery against a `K8V8` boundary. The ask is to price the *width* and
the *count*, which nobody has.

## Related

* `boundary_floor` docstring, `crates/rmlx-models/src/kv_cache/mod.rs:201–278`
* `parametric_boundary_promotion_stays_under_bf16`,
  `store_bearing_boundary_promotion_never_costs_more`
* The `k8v4` / `k8v8` retirement issue — option (2) there depends on the answer
  here.

**Comment by Pushkinist** [comment]
## Priced. The floor stays, and this issue's own byte evidence needed four corrections.

**Verdict — the count can be cut without a fidelity cost, and will not be, because it costs 11% of decode speed.** The promotion is a decode *speed-up* paid for in bytes, not a byte cost bought with quality. Cutting the tail from 8 to 4 moves perplexity by +0.28% on one architecture and −0.37% on another — no direction, inside the declared 0.5% bar — and costs 11.2–11.3% decode TPS at 32k, SEPARATED in both legs of a paired A/B. Removing the floor entirely moves perplexity by 0.019% and 0.006%. So option (2) "fewer promoted layers" is **measured and rejected**: it buys 3.2% of the cache for 11% of the speed. Option (3) "no floor at all" is measured too — no quality return at this codec and corpus — and is not taken, because a 2 048-token window on wikitext cannot see the ≥32K failure the counts were set from.

Everything below is now settable and recorded: `--kv-boundary-layers <head>,<tail>` on `serve` / `baseline` / `bench` / `eval ppl`, and `rmlx eval ppl` grew the KV flags — which required making the scorer cache-bearing first, because it forwarded each window once with **no cache at all**, so a `--kv-quant` on it would have been an inert flag.

### Corrections to the issue's own evidence

1. **`gemma-4-e2b` has p = 0 and zero power for this question.** `num_kv_shared_layers = 20` makes every tail layer a consumer that owns no cache, and its head-2 layers are sliding-attention. Three full-attention layers (4, 9, 14) carry the codec and *none* of them is promoted. Its 0.876× row is the codec alone; it says nothing about the floor. The ask names e2b as one of the two required architectures — it cannot answer. Confirmed structurally and by `store_skipped` (1 on Bonsai for `iso3_sym`, 0 on e2b) and by direct `--kv-boundary-layers 0,0` measurement.

2. **`iso3_sym`'s whole-cache win is 15.7–17.1%, not 2–4%.** The 0.962× / 0.876× figures come from a **928-token** `rmlx serve` prompt, where per-layer allocation floors and the ring's grow-step rounding are a large fraction of a small cache. At bench-scale prompts:

   | model | ctx | `iso3_sym` ÷ `none` | `k_iso3` ÷ `none` |
   |---|---|---:|---:|
   | Ternary-Bonsai-8B | 3.8k | **0.8426×** | 0.9213× |
   | Ternary-Bonsai-8B | 31.5k | **0.8294×** | 0.9147× |
   | gemma-4-e2b | 4.1k | 0.8046× | 0.9023× |
   | gemma-4-e2b | 34.4k | 0.7592× | 0.8796× |

   The gap to the 12.125 bits/value ring rate (0.758×) is 7.9% of the cache at 32k, and that gap *is* the promotion — confirmed by measuring `0,0` directly rather than deriving it.

3. **Rotor is byte-favourable, not "neutral to slightly favourable".** Its ring is 16.25 bits per value against bf16's 16.00, so replacing a rotor layer with a bf16 one gives bytes *back*: −1.29% of `rotor3_sym`'s cache at 4k, −0.64% at 32k.

4. **The in-family promotion is not free either.** The `boundary_floor` docstring said it "delivers the floor from bytes the layer already spends". It does not — the in-family target is a wider store than the base, and on Bonsai-8B at `mixed_k4g64_v4g64` the promotion is **19.80%** of the whole cache, at both 4k and 32k. What in-family buys is the comparison against the `K8V8` fallback: 0.356× whole-cache against 0.484×, worth ~27% of the cache. Cheaper, not free. Docstring and `docs/KV_QUANT.md` corrected.

   The mechanism the docs never stated: **a `K8V8`-floored layer is a bf16 layer.** `K8V8` materialises no packed store, so the fallback does not quantize the boundary layer at 8 bits — it takes the layer out of quantization and charges bf16 for it.

### Numbers, as a DB query

```sql
-- Cell 1: the linear decomposition against a direct measurement (bytes)
SELECT kv_quant, ctx_max, value, decode_config FROM observations
WHERE model = 'gemma-4-12B-it-mxfp8' AND metric = 'kv_cache_bytes'
  AND ts_utc LIKE '2026-09-04%';

-- Cell 2: fidelity
SELECT model, kv_quant, value, decode_config FROM observations
WHERE metric = 'ppl_wikitext2' AND decode_config LIKE '%ppl/scorer=cached%'
ORDER BY id;                                   -- ids 124561..124603

-- Cell 3: decode cost of the cut
SELECT id, kv_quant, value, decode_config, notes FROM observations
WHERE id BETWEEN 124553 AND 124560 ORDER BY id;
```

**Cell 1 — the decomposition is exact, not approximate.** `gemma-4-12B-it-mxfp8` (48 layers, full attention at 5/11/17/23/29/35/41/47, `num_kv_shared_layers = 0`, so F = 8 codec-bearing and p = 2 promoted). Predicting the un-floored bytes from the `2,8` run and then measuring `--kv-boundary-layers 0,0` directly:

| codec | `2,8` bytes | predicted un-floored | **measured at `0,0`** | error |
|---|---:|---:|---:|---:|
| `iso3_sym` | 393 451 520 | 389 160 960 | **389 160 960** | 0 B |
| `k_iso3` | 399 887 360 | 397 742 080 | **397 742 080** | 0 B |
| `mixed_k4g64_v4g64` | 421 450 112 | 426 492 416 | **426 492 416** | 0 B |
| `mixed_k8g64_v8g64` | 434 896 256 | 444 420 608 | **444 420 608** | 0 B |

To the byte on all four. Which also settles a side question: on a shared-KV stack the `mixed_*` family takes the `K8V8` fallback (the mirror-divert fires), which is why its rows fit the fallback formula there and the in-family one on Bonsai.

**Cell 2 — fidelity.** `rmlx eval ppl --kv-quant iso3_sym --kv-boundary-layers <h>,<t>`, wikitext-2 raw, `--ctx-window 2048 --stride 2048 --max-tokens 4096`, 4 094 scored positions. Every scored position is a single-token step through a real per-layer cache, so each NLL comes off the decode path. Deterministic — re-run reproduces every value to 10 decimals.

| model | p | `none` | `2,8` | `2,4` | `0,0` | Δ(2,4) | Δ(0,0) |
|---|---:|---:|---:|---:|---:|---:|---:|
| Ternary-Bonsai-8B | 10/36 | 10.68495 | **10.74565** | 10.77574 | 10.74358 | **+0.280%** | −0.019% |
| gemma-4-12B | 2/8 | 441.4090 | **433.7515** | 432.1616 | 433.7264 | **−0.366%** | −0.006% |

Positive control: `iso3_sym` against a bf16 cache is +0.568% on Bonsai — the same order as the effect being tested, so the nulls are nulls and not a resolution floor.

**Cell 3 — decode cost.** `scripts/perf_ab.sh --slots 8` + `--invert`, Bonsai-8B at 32k (31 553 prompt tokens), `iso3_sym`, A = `2,8`, B = `2,4`:

| leg | A median | B median | ratio | verdict | ids |
|---|---:|---:|---:|---|---|
| inverted | 12.305 | 10.930 | 0.8883 (**−11.17%**) | SEPARATED | 124553 / 124555 |
| straight | 12.307 | 10.918 | 0.8871 (**−11.29%**) | SEPARATED | 124557 / 124559 |

`kv_cache_bytes` 3 883 089 920 → 3 760 177 152 (0.9683×). Four more layers move off a bf16 SDPA over a mirror and onto the iso3 packed-ring flash kernel; that is the whole 11%. Read the other way: **the promotion is buying decode speed, and the bytes are what it pays with** — which inverts the issue's framing of it as a cost. How much an un-floored (`0,0`) run would give up is not measured here; only the 4-layer step is.

### Grade

**B.** Three cells, criteria declared before measurement (`.rmlx/analysis/443/CRITERION.md`), Cell 1 exact to the byte, Cell 3 separated in both legs. Not an A: the fidelity cell is one codec family, one corpus, and a 2 048-token window — it cannot see the ≥32K degradation the counts were originally set from, and both Gemma4 checkpoints score raw wikitext at 4 × 10² (26b at 5 × 10⁴, which is why it is not the second architecture — see the amendment in `CRITERION.md`).

### Falsifier

A long-context quality measurement — NIAH or PPL at ≥32K on a turbo/planar codec — that shows `0,0` degrading where `2,8` does not. That is the measurement the original sweep took and the one this campaign did not; it is what would convert "the floor buys nothing measurable" into "the floor buys nothing at 2k and everything at 32k".

### Invalidators

* The Cell 2 result is `iso3_sym` only. The counts were set from **turbo2** at ≥32K, a different codec family with a different loss shape.
* Cell 3 is one model at one context. The 11% is the cost of moving 4 layers between two kernels on Bonsai's geometry; it is not a constant.
* `gemma-4-12B`'s perplexity on raw wikitext is 434. Read its row as a perturbation-sensitivity measurement, not a quality one.
* A greedy token digest is **not** this measurement and must not be substituted for it: on `gemma-4-26b` at 600 tokens `mixed_k4g64_v4g64` reproduces `none` exactly while `mixed_k8g64_v8g64` diverges at index 22, and on `gemma-4-12B` at `0,0` the same pair ranks the other way. The digest reports whether a near-tie resolved the same way on one prompt; it is not ordered by fidelity.

### DB ids

* Cell 1 (bytes, gemma-4-12B): **124609–124614** at the shipped boundary, **124615–124620** at `0,0`.
* Cell 2 (PPL): **124561, 124567, 124573, 124579** (Bonsai-8B), **124585, 124591, 124597, 124603** (gemma-4-12B).
* Cell 3 (decode TPS + bytes): **124553–124556** (inverted leg), **124557–124560** (straight leg).
* The campaign's own byte rows: **124424–124460**, **124481–124492**.

One housekeeping item those last rows forced: they were filed as `decode_config = 'kv_boundary/head=2,kv_boundary/tail=8'`, an explicit spelling of the shipped default, where §3.2 says `NULL` is the engine at its defaults. Two spellings are two cells; 18 of the 61 held a cell of one that a pre-existing row would have won. `RunRecord::validate` now refuses a `decode_config` whose every term is at its default, and migration 007 moves the stored rows into the cell they always belonged in.

**Comment by Pushkinist** [comment]
### Correction to the numbers above: the Cell 2 ids moved, and one control was missing

Review of #494 found that the metric name was wrong for the cache-bearing scorer, so the Cell 2 ids I quoted are superseded. The measured values are unchanged — three separate runs reproduce every one to ten decimals — but they now live under `ppl_wikitext2_cached`, a metric of its own, because which scorer produced a perplexity changes what the number *means* and does not belong in `decode_config`. `decode_config` is the boundary's alone.

**Cell 2, final ids:**

```sql
SELECT id, model, kv_quant, value, IFNULL(decode_config,'(default boundary)')
FROM observations WHERE metric = 'ppl_wikitext2_cached' ORDER BY id;
-- 124633, 124639, 124645, 124651   Ternary-Bonsai-8B
-- 124657, 124663, 124669, 124675   gemma-4-12B
```

Cell 1 (124609–124620) and Cell 3 (124553–124560) are unchanged.

### The cross-mode control that was missing, and what it found

The review asked for a `--kv-quant none` cached-versus-cacheless pair per architecture: with a bf16 cache the two scorers should agree to floating-point noise. They do at `--ctx-window 8` — **bit-identical** on both architectures — and Qwen3 stays within ±0.003 `mean_nll` at every window from 32 up. **Gemma4 does not.**

`mean_nll`, cached minus cacheless, same corpus prefix and same scored set at each window:

| `--ctx-window` | gemma-4-12B | Ternary-Bonsai-8B |
|---:|---:|---:|
| 8 | **0.000000** | **0.000000** |
| 32 | −0.048008 | +0.003221 |
| 64 | −0.105303 | +0.001431 |
| 256 | −0.114481 | +0.000991 |
| 512 | **−0.122849** | +0.000440 |

The residual grows with attended context on a model whose sliding window is 1024, so it is not the rotating ring against the prefill SWA mask. It is a disagreement between `Gemma4Text::forward_seq_logits_all` and `Gemma4Text::forward_seq_with_cache` — two forwards that both predate this work, which giving `rmlx eval ppl` a cache-bearing mode put side by side for the first time. Filed as #495 with the sweep.

**It does not touch any conclusion in the comment above.** Every Cell 2 comparison is within the cached path at a fixed model and corpus, so the offset cancels. What it does mean is that the Gemma4 row of that table is a perturbation-sensitivity measurement and not the checkpoint's perplexity — and neither is any `ppl_*` row for a Gemma4 model already in the DB, all of which came from the cacheless forward.

### Three other things the review found, all fixed on the branch

* **A window with no scored position panicked.** `--ctx-window 4 --stride 2` on a 7-token corpus indexed past the end of the window. Reproduced, guarded before any GPU work, and the `indexing_slicing` allow now says what actually establishes the bounds.
* **A probe CSV appended to under an older header filed a `0,0` sweep in the DEFAULT cell** — silently, permanently, in an append-only table. The probe verifies the header before appending and the ingester refuses any row wider than its header.
* **An install of the boundary arriving after the first read was only a warning.** `active_kv_boundary` answers with the default when nothing is installed, so a read before the install is indistinguishable from a default run. A read now latches it and a late install is an error, checked before the store so a refused install cannot govern.

Plus: the two Python copies of the default `2,8` are gone — both ingesters read `rmlx_core::kv_boundary`, and `make check-kv-boundary-default-parity` holds the CLI help and `docs/CLI.md` against the constants.

**Comment by Pushkinist** [comment]
Resolved by #494 (main `8ccc0593`): the promotion is now settable (`--kv-boundary-layers`), priced per codec in `runs.db`, and the fidelity/throughput cells recorded in the result block above rejected the tail cut — the default stays at 2,8. The instrumentation (`rmlx eval ppl --kv-quant`, boundary in `decode_config`, default-parity gate) is what makes the remaining falsifier (the long-context fidelity arm) runnable.
